src/main/scala/cognite/spark/v1/FlexibleDataModelRelationUtils.scala

```diff
@@ -730,185 +750,139 @@ object FlexibleDataModelRelationUtils {
-    val nullable = propDef.nullable.getOrElse(true)
-    lookupFieldInRow(row, schema, propertyName, nullable) { _ =>
-      extractDirectRelation(propertyName, "Direct Node Relation", schema, defaultSpace, row)
-        .map(_.asJson)
-        .map(InstancePropertyValue.Object)
+  private def safeConvertToLong(n: BigDecimal): Long =
+    if (n.isValidLong) {
+      n.longValue
+    } else {
+      throw new IllegalArgumentException(s"'${String.valueOf(n)}' is not a valid Long")
     }
-  }
-
-  private def tryAsLong(n: Any, propertyName: String): Either[CdfSparkException, Long] = {
-    val nAsStr = String.valueOf(n)
-    val bd = BigDecimal(nAsStr)
-    if (bd.isValidLong) {
-      Right(bd.longValue)
+  private def safeConvertToInt(n: BigDecimal): Int =
+    if (n.isValidInt) {
+      n.intValue
     } else {
-      Left(new CdfSparkException(s"""Error parsing value for field '$propertyName'.
-                                    |Expecting a Long but found '$nAsStr'
-                                    |""".stripMargin))
+      throw new IllegalArgumentException(s"'${String.valueOf(n)}' is not a valid Int")
     }
-  }
-
-  private def tryAsLongSeq(ns: Seq[Any], propertyName: String): Either[CdfSparkException, Seq[Long]] =
-    Try {
-      skipNulls(ns).map { n =>
-        val bd = BigDecimal(String.valueOf(n))
-        if (bd.isValidLong) {
-          bd.longValue
-        } else {
-          throw new IllegalArgumentException(s"'${String.valueOf(n)}' is not a valid Long")
-        }
-      }
-    } match {
-      case Success(value) => Right(value)
-      case Failure(e) =>
-        val seqAsStr = ns.map(String.valueOf).mkString(",")
-        Left(new CdfSparkException(s"""Error parsing value for field '$propertyName'.
-                                       |Expecting an Array[Long] but found '[$seqAsStr]' where
-                                       |${e.getMessage}
-                                       |""".stripMargin))
+  private def safeConvertToDouble(n: BigDecimal): Double =
+    if (n.isDecimalDouble) {
+      n.doubleValue
+    } else {
+      throw new IllegalArgumentException(s"'${String.valueOf(n)}' is not a valid Double")
     }
-
-  private def tryAsInt(n: Any, propertyName: String): Either[CdfSparkException, Int] = {
-    val nAsStr = String.valueOf(n)
-    val bd = BigDecimal(nAsStr)
-    if (bd.isValidInt) {
-      Right(bd.intValue)
+  private def safeConvertToFloat(n: BigDecimal): Float =
+    if (n.isDecimalFloat) {
+      n.floatValue
     } else {
-      Left(new CdfSparkException(s"""Error parsing value for field '$propertyName'.
-                                    |Expecting an Int but found '$nAsStr'
-                                    |""".stripMargin))
+      throw new IllegalArgumentException(s"'${String.valueOf(n)}' is not a valid Float")
     }
-  }
 
-  private def tryAsIntSeq(ns: Seq[Any], propertyName: String): Either[CdfSparkException, Seq[Int]] =
+  private def tryConvertNumber[T](
+      n: Any,
```

Comment:
```suggestion
      nAsString: String,
```
`Any.toString` is bit too vague, would be better to move .toString to the caller, where hopefully a more specific type is available (or even better maybe there's a way to avoid String parsing..)

Reply:
Yeah there is a bit of "any" abuse going on, but keep in mind this is just refactoring from my part because there was so much duplicated code. The real changes related to list of direct relations are pretty small actually!

Reply:
I think you could handle it in the row.get(i) doing instead row.getAs[type](i) but it might change how errors are handled, though arguably you could just wrap that in a Try

Reply:
I was supposing that there is a reason we are doing string => bigDecimal => float instead of directly going to the final number type... but it is possible that there isn't

Reply:
If that was there before the PR then we can postpone looking deeper :eyes: , only need to check if we changed that part as with Any we get less protection by the compiler..

Reply:
Not changed. The any come from doing row.get(i)